src/Compilers/CSharp/Portable/CodeGen/EmitOperators.cs

```diff
@@ -506,6 +500,64 @@ private void EmitCondExpr(BoundExpression condition, bool sense)
             return;
         }
 
+        /// <summary>
+        /// Emits boolean expression without branching if possible (i.e., no logical operators, only comparisons).
+        /// Leaves a boolean (int32) value on the stack which conforms to sense, i.e., <c>condition == sense</c>.
+        /// </summary>
+        private bool TryEmitComparison(BoundExpression condition, bool sense)
+        {
+            RemoveNegation(ref condition, ref sense);
+
+            Debug.Assert(condition.Type.SpecialType == SpecialType.System_Boolean);
+
+            if (condition.ConstantValueOpt is { } constantValue)
+            {
+                Debug.Assert(constantValue.Discriminator == ConstantValueTypeDiscriminator.Boolean);
+                _builder.EmitBoolConstant(constantValue.BooleanValue == sense);
+                return true;
+            }
+
+            if (condition is BoundBinaryOperator binOp)
+            {
+                // Intentionally don't optimize logical operators, they need branches to short-circuit.
+                if (binOp.OperatorKind.IsComparison())
+                {
+                    EmitBinaryCondOperator(binOp, sense: sense);
+                    return true;
+                }
+            }
+            else if (condition is BoundIsOperator isOp)
+            {
+                EmitIsExpression(isOp, used: true, omitBooleanConversion: true);
+
+                // Convert to 1 or 0.
+                _builder.EmitOpCode(ILOpCode.Ldnull);
```

Comment:
I didn't follow why we're pushing `ldnull` here, but `ldc_i4_0` below (line 543). When do we need to use one versus the other?

Reply:
Never mind, I found out. `EmitIsExpression` uses `Isinst` which returns "either a null reference or an instance of that class or interface."